### PR TITLE
Auto update check

### PR DESCRIPTION
**Pull request by optizone**
Provides auto update check feature. 
Also improves way of manual update check. 

On start Cutter checks current available version. In case of available update, shows this window
(Instead of "Some newer version", there will be newest version number)
image

P.S. Prints `QNetworkReplyHttpImplPrivate::_q_startOperation was called more than once QUrl("https://api.github.com/repos/radareorg/cutter/releases/latest")` in console when checks.
It's a [known](https://bugreports.qt.io/browse/QTBUG-72463) bug. Fixed in 
5.12.2 Qt version.

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
Thank you :)
We don't want Cutter to open connection to the internet by default. This feature should only be triggered by the user in these cases:
1. Explicitly clicking a button saying "Check for available updates"
2. Opt-in to the "check for new update on start"

The opt in should be available both in the Configuration menu as well as in the Welcome dialog.

Please check here for more information: 
https://github.com/radareorg/cutter/issues/1177


Now, please follow our template for Pull Requests and describe thoroughly the changes you made. What does it add to the current "Update Button" we already have, etc

After enough information and test plan we would be able to conduct a code review :)

**Comment by optizone** [comment]
Now Cutter's users need to go to About page to check for available updates. So, if Cutter did it itself any time one starts, there wouldn't be situation when there was a new version of Cutter but users didn't know about it.

Changes:
- Added checkboxes to About page and to Welcome dialog, which allow user to choose whether he wants to use auto update check
- Added `VersionChecker` class encapsulating version check process and remove one from `AboutDialog` class

Tests:
- Run check without internet connection
- Run check with bad link
- Run a bunch of checks in a row
- Run check with 1 ms timeout
- Create a bunch of `VersionChecker` instances and run one check per one instance
- Run check normally

Demonstration:
- Screenshot of message in case of new version being attached in first comment
- Workflow of "Check for updates" button is still

image

image

Welcome dialog:
image

File: src/common/VersionChecker.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+{
+    Q_OBJECT
+public:
+    explicit VersionChecker(QObject *parent = nullptr);
+
+    /*!
+      \fn void VersionChecker::checkCurrentVersion(time_t timeoutMs)
+
+      Sends request to determine current version of Cutter.
+      If there are no respond in \a timeoutMs milliseconds, emits
```
Comment (ITAYC0HEN) [inline src/common/VersionChecker.h]: please change "there are no respond" to "there is no response"

File: src/common/VersionChecker.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      \sa checkComplete(const QString& verson, const QString& errorMsg)
+    */
+
+    void checkCurrentVersion(time_t timeoutMs);
+
+signals:
+    /*!
+      \fn VersionChecker::checkComplete(const QString& verson, const QString& errorMsg)
+
+      The signal is emited when check has been done with empty \a errorMsg string.
```
Comment (ITAYC0HEN) [inline src/common/VersionChecker.h]: "emited" -> "emitted"

"with empty" -> "with an empty"

File: src/common/VersionChecker.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      In case of error \a currVerson is empty and \a errorMsg contains description
```
Comment (ITAYC0HEN) [inline src/common/VersionChecker.h]: "In case of error \a currVerson" - >
"In case of an error, \a currVerson"

File: src/dialogs/AboutDialog.cpp [hunk 4]
```diff
@@ -86,56 +89,34 @@ void AboutDialog::on_checkForUpdatesButton_clicked()
...
+    connect(&versionChecker, &VersionChecker::checkComplete, &waitDialog, &QProgressDialog::cancel);
+    connect(&versionChecker, &VersionChecker::checkComplete,
+    [](const QString & version, const QString & error) {
+        if (error != "") {
+            QMessageBox::critical(nullptr, tr("Error!"), error);
+        } else {
+            QMessageBox mb;
+            mb.setStandardButtons(QMessageBox::Ok);
+            mb.setWindowTitle(tr("Version control"));
+            mb.setIcon(QMessageBox::Information);
```
Comment (ITAYC0HEN) [inline src/dialogs/AboutDialog.cpp]: You probably can save some lines if you'll use [this](https://doc.qt.io/qt-5/qmessagebox.html#QMessageBox-1) constructor for QMessageBox

File: src/dialogs/AboutDialog.cpp [hunk 5]
```diff
@@ -75,9 +80,7 @@ void AboutDialog::on_showPluginsButton_clicked()
 
 void AboutDialog::on_checkForUpdatesButton_clicked()
```
Comment (ITAYC0HEN) [inline src/dialogs/AboutDialog.cpp]: please document this function is you did **very well** with others. \brief, etc

File: src/dialogs/AboutDialog.cpp [hunk 6]
```diff
@@ -86,56 +89,34 @@ void AboutDialog::on_checkForUpdatesButton_clicked()
...
+        }
     });
 
+    versionChecker.checkCurrentVersion(7000);
     waitDialog.exec();
-    delete reply;
 }
 
-void AboutDialog::serveVersionCheckReply(QNetworkReply *reply)
+void AboutDialog::on_updatesCheckBox_stateChanged(int state)
```
Comment (ITAYC0HEN) [inline src/dialogs/AboutDialog.cpp]: please document this function is you did **very well** with others. \brief, etc

File: src/dialogs/AboutDialog.cpp [hunk 7]
```diff
@@ -86,56 +89,29 @@ void AboutDialog::on_checkForUpdatesButton_clicked()
...
-        mb.setText(tr("Please check your internet connection and try again."));
-        mb.exec();
+    connect(&versionChecker, &VersionChecker::checkComplete, &waitDialog, &QProgressDialog::cancel);
+    connect(&versionChecker, &VersionChecker::checkComplete,
+    [](const QString & version, const QString & error) {
+        if (error != "") {
+            QMessageBox::critical(nullptr, tr("Error!"), error);
+        } else {
+            QString text = (version == CUTTER_VERSION_FULL)
+                           ? tr("You have latest version and no need to update!")
```
Comment (ITAYC0HEN) [inline src/dialogs/AboutDialog.cpp]: "Cutter is up to date."

**Comment by a1ext** [comment]
Thanks for the PR.

I got strange behavior, I did:
- clicked `About` -> `Check for updates`
- the two message boxes appeared one after another:
image

**Comment by a1ext** [comment]
It seems like your timer got expired (you show the first message box with an error) and then the network request got finished and you shown that version is ok.

**Comment by a1ext** [comment]
One more thing, when `Check for updates on start` is turned on, then cutter appears like the following until network request is finished:

image

**Comment by optizone** [comment]
It really seems like `QNetworkAccessManager::get(QNetworkRequest)` blocks main thread, but it's not expected behavior. Maybe on windows it works like that. At least I tested only on Linux.

**Comment by optizone** [comment]
@a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.

**Comment by a1ext** [comment]
> @a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.

Yeah, sure. Which patch did you mean? This commit b735134?

**Comment by optizone** [comment]
> > @a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.
> 
> Yeah, sure. Which patch did you mean? This commit [b735134](https://github.com/radareorg/cutter/commit/b7351346d2956a2fdea7dd875e0853044a13f155)?

yes

**Comment by a1ext** [comment]
> > > @a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.
> > 
> > 
> > Yeah, sure. Which patch did you mean? This commit [b735134](https://github.com/radareorg/cutter/commit/b7351346d2956a2fdea7dd875e0853044a13f155)?
> 
> yes

No, this doesn't fix the problem.

I think it has to be implemented according to the following:
- network communication should be made in a separate thread
- the auto-check itself has to be started after the binary is chosen (e.g. after main window appears)

@ITAYC0HEN what do you think?

**Comment by ITAYC0HEN** [comment]
> network communication should be made in a separate thread

Agree - **must** be running in the background to not stop the user from working

> the auto-check itself has to be started after the binary is chosen (e.g. after main window appears)

I disagree, this check should take 2 seconds for a normal computer, and it should show when Cutter is executing. Otherwise, people would waste 5 minutes analyzing a big binary and just after they finish they will see the notification about a new version. 
They'd be happy to avoid this and update Cutter before starting the heavy analysis

**Comment by a1ext** [comment]
> > network communication should be made in a separate thread
> 
> Agree - **must** be running in the background to not stop the user from working
> 
> > the auto-check itself has to be started after the binary is chosen (e.g. after main window appears)
> 
> I disagree, this check should take 2 seconds for a normal computer, and it should show when Cutter is executing. Otherwise, people would waste 5 minutes analyzing a big binary and just after they finish they will see the notification about a new version.
> They'd be happy to avoid this and update Cutter before starting the heavy analysis

I didn't say about 5 minutes. I mean check has to run when MainWindow is going to be shown, it will be finished right after all the gadgets are filled with data.

**Comment by ITAYC0HEN** [comment]
I understood. And I have have to disagree.

Imagine some John Doe, which is a daily user of Cutter, he updates Cutter to the new release immediately when possible! :D 
Now our John open Cutter because he wants to analyze some firmware of a printer machine. John opens Cutter, selects a file, selects the default analysis and then waits 25 minutes for Cutter to finish the analysis. Then, the MainWindow is opened, the widgets are filled with data and then BOOOM. MsgBox saying "You have a new update available! It is your lucky day <3".

Now John will download the new version and would have to do everything again and to wait again 25 minutes for the new Cutter to analyze the firmware.


Instead, if the Auto Update check had happened at the start, our John was happy because he just opened Cutter and then he sees the "lucky day!" message. He will quickly download the new version and will analyze his firmware with the new Cutter.

**Comment by a1ext** [comment]
> I understood. And I have have to disagree.
> 
> Imagine some John Doe, which is a daily user of Cutter, he updates Cutter to the new release immediately when possible! :D
> Now our John open Cutter because he wants to analyze some firmware of a printer machine. John opens Cutter, selects a file, selects the default analysis and then waits 25 minutes for Cutter to finish the analysis. Then, the MainWindow is opened, the widgets are filled with data and then BOOOM. MsgBox saying "You have a new update available! It is your lucky day <3".
> 
> Now John will download the new version and would have to do everything again and to wait again 25 minutes for the new Cutter to analyze the firmware.
> 
> Instead, if the Auto Update check had happened at the start, our John was happy because he just opened Cutter and then he sees the "lucky day!" message. He will quickly download the new version and will analyze his firmware with the new Cutter.



> I understood. And I have have to disagree.
> 
> Imagine some John Doe, which is a daily user of Cutter, he updates Cutter to the new release immediately when possible! :D
> Now our John open Cutter because he wants to analyze some firmware of a printer machine. John opens Cutter, selects a file, selects the default analysis and then waits 25 minutes for Cutter to finish the analysis. Then, the MainWindow is opened, the widgets are filled with data and then BOOOM. MsgBox saying "You have a new update available! It is your lucky day <3".
> 
> Now John will download the new version and would have to do everything again and to wait again 25 minutes for the new Cutter to analyze the firmware.
> 
> Instead, if the Auto Update check had happened at the start, our John was happy because he just opened Cutter and then he sees the "lucky day!" message. He will quickly download the new version and will analyze his firmware with the new Cutter.

lol) I got your point, ok then

**Comment by optizone** [comment]
Ok. There are few things I want to say. 
First of all, `QNetworkAccessManager` works asynchronously by default, so there is no need to move it to new thread. This situation on windows **might** be because of other causes. 
Second of all, by now I can not repeat behavior @a1ext has shown. But it's because test program on windows wont even connect to the Internet. Now I investigate this issue with my windows env.

**Comment by ITAYC0HEN** [comment]
Thank you, we'll wait for further update. We won't merge until the windows
issue is fixed. Also, id like to remind that you can download Windows
artifacts from our appveyor build system :)

File: src/Main.cpp [hunk 8]
```diff
@@ -34,6 +37,32 @@ int main(int argc, char *argv[])
 
     CutterApplication a(argc, argv);
 
+    if (Config()->getAutoUpdateEnabled()) {
+        VersionChecker *versionChecker = new VersionChecker;
+        QObject::connect(versionChecker, &VersionChecker::checkComplete,
+        [](const QString & version, const QString & error) {
+            if (error == "" && version != CUTTER_VERSION_FULL) {
+                QMessageBox mb;
+                mb.setWindowTitle(QObject::tr("Version control"));
```
Comment (ITAYC0HEN) [inline src/Main.cpp]: I would be happy to see the following MsgBox:
```
There's an update available for Cutter.
Current version: XX
Latest version: YY
[ Download ] [ Don't check for updates ]  [ Cancel ]
```

Clicking "Download" will open the release link

File: src/Main.cpp [hunk 9]
```diff
@@ -34,6 +37,32 @@ int main(int argc, char *argv[])
...
+                mb.setText("<b>" + QObject::tr("Current version:") + "</b> " CUTTER_VERSION_FULL "<br/>"
+                           + "<b>" + QObject::tr("Latest version:") + "</b> " + version + "<br/><br/>"
+                           + QObject::tr("For update, please check the link:<br/>")
+                           + "<a href=\"https://github.com/radareorg/cutter/releases\">"
```
Comment (ITAYC0HEN) [inline src/Main.cpp]: If possible, I would take the user immediately to the specific release page and not to the generic page

Comment (a1ext) [reply src/Main.cpp]: Yeah, there is `html_url` in the response, it's exactly what you need.

**Comment by optizone** [comment]
I have some news. This bug appears only on Windows and Android. Code works correctly and in different thread. UI freezes appears because when first request is send Windows loads some DLL's and this blocks main thread (on my computer it is 1 second +- 60 ms and I think timings are similar on other machines). After first request is send next ones works fine. So, I see only four options:
- Load this libs before UI is launched, so user won't see freeze
- Leave it as it is (not as good as first option)
- Send request while analysis is going so UI freeze is not very noticeable 
- Do not provide this feature (not cool either)

File: src/Main.cpp (hunk 9 above)
Comment (optizone) [reply src/Main.cpp]: I think we can download directly from Cutter, so user don't need to do it himself.

Comment (ITAYC0HEN) [reply src/Main.cpp]: No, meantime let the user choose which binary of the release (32b, 64b,linux,windows, etc)

Comment (optizone) [reply src/Main.cpp]: But we know what binaries user uses, so we can help him aviod going to download page and doing all this stuff. Maybe we should give him possibility to go to this page, but button "Download" will download binaries? I did it in last commits.

**Comment by optizone** [comment]
So, I tested first option from my last comment. And I'd say that it works pretty fine. DLL load delay might even become less long (maybe compiler optimized something) and I think now windows issue is solved.

**Comment by optizone** [comment]
Btw, now Cutter will show "version control" dialog if you have latest version. It's done for debug purposes.

**Comment by a1ext** [comment]
Now its better. It would be nice if `Download` button will use save file name dialog, because we know an archive name. It will allow user to rename it before save

**Comment by optizone** [comment]
So, don't you mind this hack for Windows?

**Comment by ITAYC0HEN** [comment]
sorry I had a review ready in the last few days which I didn't submit by accident

**Pull request merged**